### DISCUSSION TITLE
Do views inherit tables RLS policies?

### DISCUSSION THREAD

**Discussion by fergusmeiklejohn** [supabase/supabase#2148] [Questions] [answered]
I'm thinking of creating a public `users` table with some columns which should be viewable only by the user. So I wonder about creating a view `user_profiles` that selects the columns from `users` that should be publicly viewable.
Is this the right approach to solving this problem? Does a view inherit the RLS on the table or tables it selects from?

thanks!

**Comment by steve-chavez** [comment, marked as answer]
Hey @fergusmeiklejohn,

There's a gotcha with views and RLS - they use the privileges of the view's owner(the role that created the view) and this is usually a superuser(which bypasses RLS) - you can see a fix for this on https://github.com/supabase/supabase/discussions/1501#discussioncomment-715670.

>  Does a view inherit the RLS on the table or tables it selects from?

RLS policies are applied to certain roles, if those roles follow RLS, then yes, the above is true.

**Reply by fergusmeiklejohn** [reply]
right thanks, so because of that it would be possible to have RLS on a table and not on a view which selects certain columns from that table? But that feels like a bit of a hack, do you know if there a more orthodox way to solve this?

**Reply by steve-chavez** [reply]
A view is most fit for hiding certain columns.

> But that feels like a bit of a hack, do you know if there a more orthodox way to solve this?

It's more of a current limitation that it's likely to be solved in a later version of postgres. More details about this [here](https://stackoverflow.com/a/33863371/4692662).

**Reply by JordyPouw** [reply]
@steve-chavez I am also trying to use a view for hiding certain columns. Of course i too get the issue with the RLS being bypassed so I change the owner of the table to `anon` because i want the table be viewable by everyone without the possibility of deleting rows.

This causes an empty table. And when i tried to create a new role and grant them select permission on the table i get a permission denied on trying to fetch the table data.

Am i doing this all wrong? or is this simply not possible?

**Comment by olirice** [comment]
Since [Postgres 15 ](https://supabase.com/blog/new-in-postgres-15#the-permission-model-is-more-secure) you can mark a view as a security_invoker and it will adopt the the permissions (including applying row level security) that are associated with the user/role making the query.

**Reply by steve-chavez** [reply]
Cross posting a snippet from https://github.com/orgs/supabase/discussions/12531#discussioncomment-5054574
